lazy/tensors/c3-compiler.rkt

```diff
@@ -0,0 +1,612 @@
+         `(scalarize
+           (flat-ext2-ρ ,f ,f-acc ,m ,n ,shape-fn ,sign
+                        (ensure-flat ,t-instrs)
+                        (ensure-flat ,u-instrs))))]
+      [(tcomp-ext1-ρ-scalar f f-acc sign tp)
+       (let ((instrs (gr-tpromise tp)))
+         `(,f ,instrs))]
+      [(tcomp-ext1-ρ f f-acc sign m shape-fn tp)
+       (let ((instrs (gr-tpromise tp)))
+         `(scalarize
+           (flat-ext1-ρ ,f ,f-acc ,m ,shape-fn ,sign
+                        (ensure-flat ,instrs))))]
+      [(tcomp-prim1-ρ f f-acc sign shape-fn tp)
+       (let ((instrs (gr-tpromise tp)))
+         `(apply-flat-ρ-fn-1 ,f ,instrs ,shape-fn))]
+      [(tcomp-prim2-ρ f f-acc sign shape-fn tp-t tp-u)
+       (let ((t-instrs (gr-tpromise tp-t))
+             (u-instrs (gr-tpromise tp-u)))
+         `(apply-flat-ρ-fn-2 ,f ,t-instrs ,u-instrs ,shape-fn))]
+      [(tcomp-prim1-∇ f f-acc sign shape-fn tp zp)
+       (let ((t-instrs (gr-tpromise tp))
+             (z-instrs (gr-tpromise zp)))
+         `(apply-flat-∇-fn-1 ,f ,t-instrs (scalarize ,z-instrs) ,shape-fn))]
+      [(tcomp-prim2-∇ fᵈ fᵈ-acc f-sign shape-fn tp-t0 tp-t1 tp-z out0 out1 i)
+       (let ((t0-instrs (gr-tpromise tp-t0))
+             (t1-instrs (gr-tpromise tp-t1))
+             (z-instrs (gr-tpromise tp-z))
+             (out-idx0 (tcomp-ds-ref-index (ext2-∇-result-res out0)))
+             (out-idx1 (tcomp-ds-ref-index (ext2-∇-result-res out1))))
+         (let ((index (if (zero? i) out-idx0 out-idx1)))
+           `(let* ([index ,index]
+                   [v (data-segment-ref index)])
+              (cond
+                ((eqv? v 'uncalculated)
+                 (prim2-∇-forcer! ,fᵈ ,fᵈ-acc ,f-sign ,shape-fn
+                                  ,t0-instrs ,t1-instrs
+                                  ,z-instrs ,out-idx0 ,out-idx1)
+                 (data-segment-ref index))
+                (else v)))))]
+      [(tcomp-reshape s tp)
+       (let ((instrs (gr-tpromise tp)))
+         `(flat ',s
+                (flat-store ,instrs)
+                (flat-offset ,instrs)))]
+      [(tcomp-let lhs rhs body)
+       (let ((rhs-instrs (gr-tpromise rhs))
+             (body-instrs (gr-tpromise body)))
+         `(let ((,lhs ,rhs-instrs))
+            ,body-instrs))]
+      [(tcomp-var name) name]
+      [(tcomp-ds-ref index) `(data-segment-ref ,index)])))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Composing Compiler Passes
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define print-compiler? (make-parameter #f))
+(define display-compiler-trace
+  (λ (title value)
+    (when (or (equal? (print-compiler?) #t)
+              (and (list? (print-compiler?)) (member title (print-compiler?))))
+      (printf "~a:~n" title)
+      (displayln "--------------")
+      (pretty-print value)
+      (displayln ""))))
+(define cache
+  (make-parameter (make-hash)))
+(define compile-tensor
+  (λ (t)
+    (display-compiler-trace 'Source-Tensor t)
```

Comment:
I like to see top level compiler codes look compositional:
```
(pass1 args ... 
   (pass2 args ...
       (pass3 args ... and so on
           final-code-handler-k)))
``` 
Here, each pass-n returns a continuation expecting values from the previous pass, and expects a continuation for the next pass to pass its values to. 

That way you can also restrict the tracing in each of the individual pass-n functions